dashboard/app/controllers/featured_projects_controller.rb

```diff
@@ -1,23 +1,19 @@
 class FeaturedProjectsController < ApplicationController
-  def create
-    unless current_user.try(:permission?, UserPermission::PROJECT_VALIDATOR)
-      _, channel_id = storage_decrypt_channel_id(featured_project_params[:project_id])
-      @featured_project = FeaturedProject.create({storage_app_id: channel_id, who_featured_user_id: current_user.id})
-      @featured_project.save!
-    end
-  end
+  authorize_resource
 
-  def destroy_by_project_id
-    unless current_user.try(:permission?, UserPermission::PROJECT_VALIDATOR)
-      _, channel_id = storage_decrypt_channel_id(params[:project_id])
-      @featured_project = FeaturedProject.find_by_storage_app_id(channel_id)
-      @featured_project.destroy
-    end
+  def feature
+    _, channel_id = storage_decrypt_channel_id(params[:project_id])
+    return render_404 unless channel_id
+    @featured_project = FeaturedProject.find_or_create_by(storage_app_id: channel_id)
+    @featured_project.update_attribute(:unfeatured_at, nil)
+    @featured_project.update_attribute(:featured_at, DateTime.now)
+    @featured_project.save!
   end
 
-  private
-
-  def featured_project_params
-    params.require(:featured_project).permit(:storage_app_id, :project_id, :who_featured_user_id)
+  def unfeature
+    _, channel_id = storage_decrypt_channel_id(params[:project_id])
+    return render_404 unless channel_id
+    @featured_project = FeaturedProject.find_by storage_app_id: channel_id
+    @featured_project.update_attribute(:unfeatured_at, DateTime.now)
```

Comment:
`find_by` will return nil if the record is not found, and then the next line will blow up accessing nil. What *should* we do if there is nothing to unfeature? If we want to fail and log to honeybadger, use `find_by!` to get a `RecordNotFound` error. If we want to return bad request or something we'll need more code.

Don't use [update_attribute](https://apidock.com/rails/ActiveRecord/Persistence/update_attribute) unless you specifically want it because it has side effects (or rather lack of normal expected side effects), notably it skips validations and callbacks. While that's probably not relevant here, unless we specifically want it, it's better practice to do something like:

```ruby
@featured_project.update! unfeatured_at: DateTime.now
```

A question on idempotence: what should the time be if we unfeature a project that is already unfeatured? My guess is we don't care in which case this is fine, but note it will always have the latest time not the first time so it might be inaccurate. Alternatively, only update `unfeatured_at` if it's nil.

Reply:
We don't what the time is for unfeaturing a project that was previously unfeatured. `unfeatured_at` is more like "the time when this project was most recently unfeatured"; ant previous unfeaturing is irrelevant.